### PR TITLE
Adding NCHW FP16 SIMD kernel

### PR DESCRIPTION
**Pull request by SundarRajan28**

### CODE REVIEW

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 1]
```diff
@@ -793,15 +793,84 @@ MasterGraph::copy_out_tensor(void *out_ptr, RocalTensorFormat format, float mult
...
+                                fG = _mm256_cvtepi32_ps(_mm256_shuffle_epi8(pix0, mask_G));
+                                fR = _mm256_cvtepi32_ps(_mm256_shuffle_epi8(pix0, mask_B));
+                                fB = _mm256_fmadd_ps(fB, pmul0, padd0);
+                                fG = _mm256_fmadd_ps(fG, pmul1, padd1);
+                                fR = _mm256_fmadd_ps(fR, pmul2, padd2);
+                                _mm256_storeu_ps(tempB, fB);
+                                _mm256_storeu_ps(tempG, fG);
+                                _mm256_storeu_ps(tempR, fR);
+                                for(int cnt = 0; cnt < 8; cnt++)
+                                {
```
Comment (rrawther) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:850]: move the { to the previous line to follow google coding style

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 2]
```diff
@@ -793,15 +793,84 @@ MasterGraph::copy_out_tensor(void *out_ptr, RocalTensorFormat format, float mult
...
+                                    B_buf_16[cnt] = (half) tempB[cnt];
```
Comment (rrawther) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:851]: please do the conversion from float to half in SSE using _mm_cvtps_ph() or similar instruction and get rid of loop here

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 3]
```diff
@@ -793,15 +793,84 @@ MasterGraph::copy_out_tensor(void *out_ptr, RocalTensorFormat format, float mult
...
+                            for (; i < alignedLength; i += 8) {
+                                __m256i pix0 = _mm256_loadu_si256((const __m256i *) in_buffer);
+                                pix0 = _mm256_permutevar8x32_epi32(pix0, _mm256_setr_epi32(0, 1, 2, 3, 3, 4, 5, 6));
+                                fB = _mm256_cvtepi32_ps(_mm256_shuffle_epi8(pix0, mask_R));
+                                fG = _mm256_cvtepi32_ps(_mm256_shuffle_epi8(pix0, mask_G));
+                                fR = _mm256_cvtepi32_ps(_mm256_shuffle_epi8(pix0, mask_B));
+                                fB = _mm256_fmadd_ps(fB, pmul0, padd0);
+                                fG = _mm256_fmadd_ps(fG, pmul1, padd1);
+                                fR = _mm256_fmadd_ps(fR, pmul2, padd2);
+                                _mm256_storeu_ps(tempB, fB);
```
Comment (rrawther) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:846]: get rid of temp buffers and store the result after conversion

**Comment by kiritigowda** [comment]
@SundarRajan28 any updates on the review?

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 2 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp]: Created a new commit adding _mm_cvtps_ph() for conversion from float to half

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 3 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp]: Created a new commit removing the temp buffer usage and used registers for intermediate results

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 4]
```diff
@@ -843,15 +843,12 @@ MasterGraph::copy_out_tensor(void *out_ptr, RocalTensorFormat format, float mult
...
-                                _mm256_storeu_ps(tempB, fB);
-                                _mm256_storeu_ps(tempG, fG);
-                                _mm256_storeu_ps(tempR, fR);
-                                for(int cnt = 0; cnt < 8; cnt++)
-                                {
-                                    B_buf_16[cnt] = (half) tempB[cnt];
-                                    G_buf_16[cnt] = (half) tempG[cnt];
-                                    R_buf_16[cnt] = (half) tempR[cnt];
-                                }
+                                tempB = _mm256_cvtps_ph(fB, _MM_FROUND_NO_EXC);
```
Comment (rrawther) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:846]: probably you need to specify rounding mode here (round towards zero)

Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp]: Used _MM_FROUND_TO_ZERO as rounding mode for rounding towards zero

**Review by rrawther — APPROVED** [review]

**Pull request merged**